lib/vsc/mympirun/pmi/sched.py

```diff
@@ -0,0 +1,180 @@
+#
+# vsc-mympirun is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with vsc-mympirun.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Base PMI Sched class, all actual classes should inherit from this one
+
+The role of the Sched class is mainly to construct the correct sched-specific PMI call
+"""
+
+import os
+from vsc.utils.fancylogger import getLogger
+from vsc.mympirun.common import SchedBase
+from vsc.utils.run import run_file, async_to_stdout
+
+# simple intermediate representation of the job resources
+JOB_INFO = {
+    'tnodes': None,  # total number of nodes
+    'ncores': None,  # number of cores per node
+    'nranks': None, # number of MPI ranks per node
+    'nmem': None,  # memory per node
+    'ngpus': None,  # number of gpus per node: None means no GPUs present, 0 means don't use GPUs
+}
+
+class Sched(SchedBase):
+    LAUNCHER = None
+
+    def __init__(self, options=None, **kwargs):
+        if not hasattr(self, 'log'):
+            self.log = getLogger(self.__class__.__name__)
+        if not hasattr(self, 'options'):
+            self.options = options
+
+        self.envs = []  # list of enviroment variable names that are modified
+
+        super(Sched, self).__init__(**kwargs)
+
+    def set_env(self, key, value, keep=False):
+        """
+        Set os.environ and track variable
+        If keep is True, don't override existing value
+        """
+        current = os.environ.get(key)
+        if keep and current is not None:
+            self.log.debug("Keeping existing environment variable %s with value %s (ignore value %s)",
+                           key, current, value)
+        else:
+            os.environ[key] = str(value)
+            self.envs.append(key)
+            self.log.debug("Set environment variable %s: %s", key, value)
+
+    def pmicmd(self):
+        """
+        Return generated pmi command (as list) and the run function
+        envs is list of variable names that is modified
+        """
+        pmicmd = [self.LAUNCHER]
+
+        # the mpi calls should only set environment variables
+        for name in ['tune', 'pmi', 'debug']:
+            method_name = 'mpi_' + name
+            getattr(self, method_name)()
+            self.log.debug("Calling %s", method_name)
+
+        for name in ['sched', 'sizing', 'environment', 'mpi', 'debug']:
+            args = getattr(self, 'pmicmd_' + name)()
+            self.log.debug("Generated pmicmd %s arguments %s", name, args)
+            pmicmd.extend(args)
+
+        run_function, run_function_args = self.run_function()
+        pmicmd.extend(run_function_args)
+
+        pmicmd.extend(['--' + x for x in getattr(self.options, 'pass', [])])  # .pass gives syntax error?
```

Comment:
I still think you should rename `.pass` to something else rather than hack around it :)

Reply:
i don't. pass is a fitting name; i'll work around the python quirks